component_sdk/python/kfp_component/core/_display.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import json
+import threading
+import logging
+
+_OUTPUT_PATH = os.environ.get('KFP_UI_METADATA_PATH', '/mlpipeline-ui-metadata.json')
+_OUTPUT_FILE_LOCK = threading.Lock()
+
+def display(obj):
+    """Display an object to KFP UI.
+    """
+    obj_dir = dir(obj)
+    if '_repr_html_' in obj_dir:
+        display_html(obj)
+
+    if '_repr_kfpmetadata_' in obj_dir:
+        display_kfpmetadata(obj)
+
+def display_html(obj):
+    """Display html representation to KFP UI.
+    """
+    if '_repr_html_' not in dir(obj):
+        raise ValueError('_repr_html_ function is not present.')
+    html = obj._repr_html_()
+    _output_ui_metadata({
+        'type': 'web-app',
+        'html': html
+    })
+
+def display_kfpmetadata(obj):
+    """Display from KFP UI metadata
+    """
+    if '_repr_kfpmetadata_' not in dir(obj):
+        raise ValueError('_repr_kfpmetadata_ function is not present.')
+    kfp_metadata = obj._repr_kfpmetadata_()
+    _output_ui_metadata(kfp_metadata)
+
+def _output_ui_metadata(output):
+    logging.info('Dumping metadata: {}'.format(output))
+    with _OUTPUT_FILE_LOCK:
```

Comment:
I suggest using a lock timeout to avoid unbounded blocking call to the lock acquire.

Reply:
Python 2.7 doesn't support lock with timeout. It will be some work to support it. I think the long term plan was to make this file as a recordio file to support appending which we don't have to lock the file in POSIX systems.

Reply:
SG